submissions/dashakim/a-tiny-world2/index.js

```diff
@@ -0,0 +1,67 @@
+class Inhabitant {
+  constructor(name, gender) {
+    this.resident = this.constructor.name;
+    this.name = name;
+    this.gender = gender;
+  }
+  toSring() {
+    return Object.entries(this)
+      .map(([prop, value]) => `${prop}: <b>${value}</b>`)
+      .join("; ");
+  }
+}
+const SayCatMixin = superclass =>
+  class extends superclass {
+    constructor(name, gender, saying) {
+      super(name, gender, saying);
+      this.saying = "Meow";
+    }
+  };
+class Human extends Inhabitant {
+  constructor(name, gender, saying) {
+    super(name, gender);
+    this.hands = 2;
+    this.legs = 2;
+    this.saying = saying;
+  }
+}
+class Man extends Human {
+  constructor(name, saying) {
+    super(name, "male", saying);
+  }
+}
+class Woman extends Human {
+  constructor(name, saying) {
+    super(name, "female", saying);
+  }
+}
+class Animal extends Inhabitant {
+  constructor(name, gender, saying) {
+    super(name, gender);
+    this.paws = 4;
+    this.saying = saying;
+  }
+}
+class Dog extends Animal {
+  constructor(name, gender, saying) {
+    super(name, gender, saying);
+  }
+}
+class Cat extends SayCatMixin(Animal) {
+  constructor(name, gender) {
+    super(name, gender);
+  }
+}
+class CatWoman extends SayCatMixin(Woman) {
+  constructor(name, saying) {
+    super(name, saying);
+  }
+}
+let inhabitants = [
+  new Man("Jack", "I hope you're having a wonderful day"),
+  new Woman("Sara", "Hey Jack, good to see you!"),
+  new Dog("Archi", "male", "Woof-woof!"),
+  new Cat("Felix", "male"),
+  new CatWoman("Selina")
```

Comment:
Two things look a bit weird here.

1. `Dog` requires specifying its saying explicitly while `Cat` does not. This is a somewhat inconsistent API. Imagine we have to extend the taxonomy with other animals. Should our user remember what classes require saying to be specified and what do not? There must be a simple rule. E.g. humans require a saying, others do not. Cat-woman can have a default saying if neither specified.
Also note that `CatWoman` constructor provides for a `saying` argument. What will happen if a user creates CatWoman with an individual saying? Perhaps the output will be something the user didn't expect.

2. The spec says that output log for each inhabitant must contain her/his/its species. We can understand whether a human is a Woman or a Man by her/his gender. This leads us to having the `species` property instead of ambiguous `resident`.

OOP is about well designed, maintainable, and scalable data structure.

Please let me know if you're gonna take any of the two challenges above.

Reply:
Good point! Thank you! I will change this things.